### PR TITLE
client,cmd: support removing components and snaps with components

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Adding support and tests to remove components using `snap` command.

### CODE REVIEW

**Review by andrewphelpsj — APPROVED** [review]
thanks!

File: cmd/snap/cmd_snap_op.go [hunk 1]
```diff
@@ -138,10 +138,68 @@ type cmdRemove struct {
...
+	}
+}
+
 func (x *cmdRemove) removeOne(opts *client.SnapOptions) error {
-	name := string(x.Positional.Snaps[0])
+	arg := string(x.Positional.Snaps[0])
 
-	changeID, err := x.client.Remove(name, opts)
+	name, comps := splitSnapAndComponents(arg)
+	changeID, err := x.client.Remove(name, comps, opts)
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:202]: if comps is not empty this remove only the components and not the snap right?

File: cmd/snap/cmd_snap_op.go [hunk 2]
```diff
@@ -622,12 +681,18 @@ func splitSnapAndComponents(name string) (string, []string) {
...
 	allComps := make(map[string][]string, len(names))
 	for _, name := range names {
 		snap, comps := splitSnapAndComponents(name)
-		snaps = append(snaps, snap)
+		// When installing we implicitly want to install the snap when
+		// we have specified also components, but when removing we
+		// actually want to remove only components if any of them have
+		// been specified.
+		if forInstall || len(comps) == 0 {
+			snaps = append(snaps, snap)
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:694]: but this is a bit different, because one uses the /{name} endpoint?

File: cmd/snap/cmd_snap_op.go [hunk 3]
```diff
@@ -151,24 +209,38 @@ func (x *cmdRemove) removeOne(opts *client.SnapOptions) error {
...
+	showRemoved(expectedRemovedSnaps, map[string][]string{name: comps}, snapToRev, changes)
+
 	return nil
 }
 
 func (x *cmdRemove) removeMany(opts *client.SnapOptions) error {
 	names := installedSnapNames(x.Positional.Snaps)
-	changeID, err := x.client.RemoveMany(names, opts)
+
+	forInstall := false
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:241]: nitpick: can be a const?

File: cmd/snap/cmd_snap_op.go [hunk 4]
```diff
@@ -712,7 +777,8 @@ func (x *cmdInstall) installMany(names []string, opts *client.SnapOptions) error
 			return errors.New(i18n.G("cannot specify mode for multiple store snaps (only for one store snap or several local ones)"))
 		}
 
-		names, comps := snapsAndComponentsFromNames(names)
+		forInstall := true
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:780]: same

**Review by pedronis — COMMENTED** [review]
lgtm, just some nitpicks and double checking questions

File: cmd/snap/cmd_snap_op.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply cmd/snap/cmd_snap_op.go:204]: Yes, that is right, I've added a comment to clarify.

File: cmd/snap/cmd_snap_op.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply cmd/snap/cmd_snap_op.go]: Sure , done

File: cmd/snap/cmd_snap_op.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply cmd/snap/cmd_snap_op.go]: changed

File: cmd/snap/cmd_snap_op.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply cmd/snap/cmd_snap_op.go:696]: This function is called by the *Many methods in fact so we have a list of snaps and a map of components per snap.

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks, comments addressed now. I've also squashed the commit with the changes.

File: client/snap_op.go [hunk 5]
```diff
@@ -148,12 +148,12 @@ func (client *Client) InstallMany(names []string, components map[string][]string
 }
 
 // Remove removes the snap with the given name.
```
Comment (pedronis) [inline client/snap_op.go:150]: probably worth expanding the doc comment here as well.

File: client/snap_op.go [hunk 6]
```diff
@@ -148,12 +148,12 @@ func (client *Client) InstallMany(names []string, components map[string][]string
...
-func (client *Client) Remove(name string, options *SnapOptions) (changeID string, err error) {
-	return client.doSnapAction("remove", name, nil, options)
+func (client *Client) Remove(name string, components []string, options *SnapOptions) (changeID string, err error) {
+	return client.doSnapAction("remove", name, components, options)
 }
 
-func (client *Client) RemoveMany(names []string, options *SnapOptions) (changeID string, err error) {
-	return client.doMultiSnapAction("remove", names, nil, options)
+func (client *Client) RemoveMany(names []string, components map[string][]string, options *SnapOptions) (changeID string, err error) {
```
Comment (pedronis) [inline client/snap_op.go:155]: and adding one here

**Review by pedronis — APPROVED** [review]
lgtm, but probably worth adding/expanding doc comments in client

**Pull request merged**